File: engine/aurion/license/ed25519.py
"""Pure-Python Ed25519 (RFC 8032) — no third-party dependency.

Used for license-key signatures: the engine ships ONLY the public key, so a
public repository (or a curious customer) can verify keys but can never mint
them. Signing needs the 32-byte private seed, held exclusively by the owner
(keyserver env var / owner CLI).

API:
    publickey(seed32) -> 32 bytes
    sign(seed32, msg) -> 64 bytes
    verify(sig64, msg, pub32) -> bool
"""

from __future__ import annotations

import hashlib

_b = 256
_q = 2**255 - 19
_l = 2**252 + 27742317777372353535851937790883648493


def _H(m: bytes) -> bytes:
    return hashlib.sha512(m).digest()


def _inv(x: int) -> int:
    return pow(x, _q - 2, _q)


_d = (-121665 * _inv(121666)) % _q
_I = pow(2, (_q - 1) // 4, _q)


def _xrecover(y: int) -> int:
    xx = (y * y - 1) * _inv(_d * y * y + 1) % _q
    x = pow(xx, (_q + 3) // 8, _q)
    if (x * x - xx) % _q != 0:
        x = (x * _I) % _q
    if x % 2 != 0:
        x = _q - x
    return x


_By = (4 * _inv(5)) % _q
_Bx = _xrecover(_By)
_B = (_Bx, _By)


def _edwards(P: tuple[int, int], Q: tuple[int, int]) -> tuple[int, int]:
    (x1, y1) = P
    (x2, y2) = Q
    t = (_d * x1 * x2 * y1 * y2) % _q
    x3 = (x1 * y2 + x2 * y1) * _inv(1 + t) % _q
    y3 = (y1 * y2 + x1 * x2) * _inv(1 - t) % _q
    return (x3, y3)


def _scalarmult(P: tuple[int, int], e: int) -> tuple[int, int]:
    if e == 0:
        return (0, 1)
    Q = _scalarmult(P, e >> 1)
    Q = _edwards(Q, Q)
    if e & 1:
        Q = _edwards(Q, P)
    return Q


def _encodeint(y: int) -> bytes:
    return int(y).to_bytes(32, "little")


def _encodepoint(P: tuple[int, int]) -> bytes:
    (x, y) = P
    out = bytearray(_encodeint(y))
    out[31] |= (x & 1) << 7
    return bytes(out)


def _decodeint(s: bytes) -> int:
    return int.from_bytes(s, "little")


def _bit(h: bytes, i: int) -> int:
    return (h[i // 8] >> (i % 8)) & 1


def _prune(h: bytes) -> int:
    a = 2 ** (_b - 2)
    for i in range(3, _b - 2):
        a += 2**i * _bit(h, i)
    return a


def _isoncurve(P: tuple[int, int]) -> bool:
    (x, y) = P
    return (-x * x + y * y - 1 - _d * x * x * y * y) % _q == 0


def _decodepoint(s: bytes) -> tuple[int, int]:
    if len(s) != 32:
        raise ValueError("point must be 32 bytes")
    y = _decodeint(s) & ((1 << 255) - 1)
    x = _xrecover(y)
    if (x & 1) != (s[31] >> 7):
        x = _q - x
    if not _isoncurve((x, y)):
        raise ValueError("point is not on the curve")
    return (x, y)


def _Hint(m: bytes) -> int:
    return int.from_bytes(_H(m), "little")


def publickey(seed: bytes) -> bytes:
    if len(seed) != 32:
        raise ValueError("seed must be 32 bytes")
    h = _H(seed)
    a = _prune(h)
    return _encodepoint(_scalarmult(_B, a))


def sign(seed: bytes, msg: bytes) -> bytes:
    if len(seed) != 32:
        raise ValueError("seed must be 32 bytes")
    h = _H(seed)
    a = _prune(h)
    A = _encodepoint(_scalarmult(_B, a))
    r = _Hint(h[32:64] + msg) % _l
    R = _scalarmult(_B, r)
    S = (r + _Hint(_encodepoint(R) + A + msg) * a) % _l
    return _encodepoint(R) + _encodeint(S)


def verify(sig: bytes, msg: bytes, pub: bytes) -> bool:
    try:
        if len(sig) != 64 or len(pub) != 32:
            return False
        R = _decodepoint(sig[:32])
        A = _decodepoint(pub)
        S = _decodeint(sig[32:])
        if S >= _l:
            return False
        h = _Hint(sig[:32] + pub + msg) % _l
        return _scalarmult(_B, S) == _edwards(R, _scalarmult(A, h))
    except (ValueError, IndexError):
        return False


if __name__ == "__main__":  # keygen:  python -m aurion.license.ed25519
    import os as _os

    _seed = _os.urandom(32)
    print("private seed (KEEP SECRET — keyserver .env AURION_KEY_PRIVATE_HEX):", _seed.hex())
    print("public key   (ships in material.py ED25519_PUBLIC_HEX):           ", publickey(_seed).hex())

File: keyserver/src/ed25519.js
"use strict";
// Pure-JS Ed25519 (RFC 8032) with BigInt — byte-exact port of
// engine/aurion/license/ed25519.py. Verified in test/key-parity.mjs against
// the official RFC test vectors, so Node and Python always agree.
//
// Only THIS server (and the owner CLI) may hold a private seed.

const crypto = require("crypto");

const q = 2n ** 255n - 19n;
const l = 2n ** 252n + 27742317777372353535851937790883648493n;

const H = (m) => crypto.createHash("sha512").update(m).digest();
const mod = (v) => ((v % q) + q) % q;

function modPow(b, e, m) {
  let r = 1n;
  b = ((b % m) + m) % m;
  while (e > 0n) {
    if (e & 1n) r = (r * b) % m;
    b = (b * b) % m;
    e >>= 1n;
  }
  return r;
}

const inv = (x) => modPow(x, q - 2n, q);
const d = mod(-121665n * inv(121666n));
const I = modPow(2n, (q - 1n) / 4n, q);

function xrecover(y) {
  const xx = mod((y * y - 1n) * inv(mod(d * y * y + 1n)));
  let x = modPow(xx, (q + 3n) / 8n, q);
  if (mod(x * x - xx) !== 0n) x = mod(x * I);
  if (x % 2n !== 0n) x = q - x;
  return x;
}

const By = mod(4n * inv(5n));
const Bx = xrecover(By);
const B = [Bx, By];

function edwards(P, Q) {
  const [x1, y1] = P;
  const [x2, y2] = Q;
  const t = mod(d * x1 * x2 * y1 * y2);
  const x3 = mod((x1 * y2 + x2 * y1) * inv(mod(1n + t)));
  const y3 = mod((y1 * y2 + x1 * x2) * inv(mod(1n - t)));
  return [x3, y3];
}

function scalarmult(P, e) {
  if (e === 0n) return [0n, 1n];
  let Q = scalarmult(P, e >> 1n);
  Q = edwards(Q, Q);
  if (e & 1n) Q = edwards(Q, P);
  return Q;
}

function encodeint(y) {
  const out = Buffer.alloc(32);
  let v = BigInt(y);
  for (let i = 0; i < 32; i++) {
    out[i] = Number(v & 0xffn);
    v >>= 8n;
  }
  return out;
}

function encodepoint(P) {
  const [x, y] = P;
  const out = encodeint(y);
  out[31] |= Number(x & 1n) << 7;
  return out;
}

const decodeint = (s) => {
  let v = 0n;
  for (let i = s.length - 1; i >= 0; i--) v = (v << 8n) | BigInt(s[i]);
  return v;
};

const bitAt = (h, i) => Number((h[i >> 3] >> (i % 8)) & 1);

function prune(h) {
  let a = 2n ** 254n;
  for (let i = 3; i < 254; i++) a += 2n ** BigInt(i) * BigInt(bitAt(h, i));
  return a;
}

function isoncurve(P) {
  const [x, y] = P;
  return mod(-x * x + y * y - 1n - d * x * x * y * y) === 0n;
}

function decodepoint(s) {
  if (s.length !== 32) throw new Error("point must be 32 bytes");
  const y = decodeint(s) & ((1n << 255n) - 1n);
  let x = xrecover(y);
  if ((x & 1n) !== BigInt(s[31] >> 7)) x = q - x;
  if (!isoncurve([x, y])) throw new Error("point is not on the curve");
  return [x, y];
}

const Hint = (m) => decodeint(H(m));

function publickey(seed) {
  if (seed.length !== 32) throw new Error("seed must be 32 bytes");
  const h = H(seed);
  const a = prune(h);
  return encodepoint(scalarmult(B, a));
}

function sign(seed, msg) {
  if (seed.length !== 32) throw new Error("seed must be 32 bytes");
  const h = H(seed);
  const a = prune(h);
  const A = encodepoint(scalarmult(B, a));
  const r = Hint(Buffer.concat([h.subarray(32, 64), msg])) % l;
  const R = scalarmult(B, r);
  const S = (r + Hint(Buffer.concat([encodepoint(R), A, msg])) * a) % l;
  return Buffer.concat([encodepoint(R), encodeint(S)]);
}

function verify(sig, msg, pub) {
  try {
    if (sig.length !== 64 || pub.length !== 32) return false;
    const R = decodepoint(sig.subarray(0, 32));
    const A = decodepoint(pub);
    const S = decodeint(sig.subarray(32));
    if (S >= l) return false;
    const h = Hint(Buffer.concat([sig.subarray(0, 32), pub, msg])) % l;
    const left = scalarmult(B, S);
    const right = edwards(R, scalarmult(A, h));
    return left[0] === right[0] && left[1] === right[1];
  } catch {
    return false;
  }
}

module.exports = { publickey, sign, verify };
